Use augur subsample

victorlin opened the pull request:
> [!NOTE]
> Needs changes from https://github.com/nextstrain/augur/pull/2038

## Description of proposed changes

The previous subsampling implementation was fixed to a two-sample recent+background split with some hardcoded parameters. Replacing it with augur subsample allows for more flexible configuration.

In Snakemake, implementation is mostly copied from pathogen repos that have switched over to augur subsample. One notable difference is that the combine_samples rule must stay to handle output from the enrich_antibody_escape rule.

This is a breaking change and the old configuration will no longer work.

## Related issue(s)

Closes #101

## Checklist

- [x] Update changelog
- [x] Draft necessary changes on NW-PaGe fork: 0582db58741c04da8c99bd0a4c0027f78d4e462b
- [x] Manually tested that validation fails with proximal sample entry
- [ ] Merge #139 and https://github.com/nextstrain/augur/pull/2038
- [ ] Checks pass

<details>
<summary>old implementations</summary>

- YAML anchors and aliases: b0d6728f16c8a46ae8cfcd5113c19174cafb0df7
- matrix: d24a8d10427493cb17dd4c980d048f470cdfac29
- dataset-first glob syntax 1: 5741f0f28a9faa636037f23e36e9949a356c5d95
- option-first glob syntax 1: bff7e53ba1691aff2a10d89681d9075326594dc2
- option-first glob syntax 2: 1df4e05550cea101cec8bd9b82d471c593e0c61e
- dataset-first glob syntax 2: 397850dda581f3fd0dee183f6078c8de6f590f2a

</details>

joverlee521 commented on config/configfile.yaml:
```diff
@@ -6,46 +55,65 @@ builds_to_run: ["genome", "G", "F"]
+  genome/all-time:
+    samples:
+      all-time:
+        <<: [*subsample_genome, *subsample_all-time, *subsample_defaults]
```
In general, I'm not sure this pattern of YAML anchors/aliases is very user friendly. If we do use anchors/aliases, then we should probably disable them in the dumped config so that users can at least see the fully expanded config in `results/run_config.yaml` (see suggested workaround in https://github.com/yaml/pyyaml/issues/103).

I wonder if this can use the config wildcards pattern that @jameshadfield used in [avian-flu](https://github.com/nextstrain/avian-flu/blob/f1657094ebf9da39660a7f4ba6a5ddf51b6a9c3e/segment-focused/config.yaml#L71). These could be expanded at start of Snakemake (or whatever comes of discussion in https://github.com/nextstrain/public/issues/23).

jameshadfield replied:
> If we do use anchors/aliases, then we should probably disable them in the dumped config 

Yes, agreed. (I'd go further: we should write out small-multiples with lots of duplication, but that's beyond this PR.)

> I wonder if this can use the config wildcards pattern that @jameshadfield used in [avian-flu](https://github.com/nextstrain/avian-flu/blob/f1657094ebf9da39660a7f4ba6a5ddf51b6a9c3e/segment-focused/config.yaml#L71).


Across the 9 builds almost all of the difference is the subsampling parameters, so at some level it is going to be either complex or verbose. I don't think glob-like syntax will help here with the structure as it, but it could if you moved towards something more like:

```yaml
subsample:
  samples:
    min_length:
      "genome/*": 10_000
      "G/*": 600
      "F/*": 1_200
    group_by: "year country"
    min_coverage: 0.3
    resolutions:
      "*/all-time": {"min_date": "1795-01-01"}
      "*/6y": {"min_date": "6Y", "background_min_date": "1975-01-01"}
      "*/3y": {"min_date": "3Y", "background_min_date": "1975-01-01"}
```

j23414 replied:
For my ability to follow this thread, the reason against using YAML anchors and aliases:
* not sure if the pattern of YAML anchors/aliases is very user friendly (for which users? A google search is returning several articles recommending YAML anchors ([example](https://medium.com/@kinghuang/docker-compose-anchors-aliases-extensions-a1e4105d70bd)) to avoid YAML duplication, but maybe these are not our user group or feel free to add comments from WA-DOH/others etc.)
* requires resolving the anchors explicitly in the dumped config (similar lift as `config pre-processing to translate into augur subsample-ready config`)

Feel free to add to the above list, mostly looking for a summary statement

jameshadfield replied:
> Anchors are resolved automatically, i.e. yaml.dump(yaml.load(...)) will not preserve anchors.

I got this wrong (see https://github.com/nextstrain/shared/pull/62), at least partially. Anchors of primitive types won't be preserved, but anchors of more complex types are preserved. E.g. the following YAML

```yaml
foo: &foo
  a: 1
  b: 2
bar: *foo

x: &x 42
y: *x
```

when roundtripped through `yaml.dump(yaml.load(...))` will keep one of the anchors and drop the other:

```yaml
bar: &id001
  a: 1
  b: 2
foo: *id001
x: 42
y: 42
```

victorlin replied:
I had YAML anchors+aliases mostly working in 466d7cbf6d36de88f3525651e74b09b096646246, but it was pretty ugly.

The logic is more readable as Python code that generates each small-multiple subsample config from a more concise workflow config (but still flexible), implemented in 1ccf820655e23c49bd1512495698c6bfd0fadb57.

victorlin replied:
Updated to a new version of "glob syntax" in e54f105fb9dcb8daa25443edec71751452708a3c:

https://github.com/nextstrain/rsv/blob/e54f105fb9dcb8daa25443edec71751452708a3c/config/configfile.yaml#L31-L100

victorlin replied:
Pushed up another version of glob syntax in bff7e53ba1691aff2a10d89681d9075326594dc2. This is the same structure as avian-flu with some additional features.

https://github.com/nextstrain/rsv/blob/08eb257a71248129f5f98eff202a68bb213e941f/config/configfile.yaml#L69-L77

Also added a draft of the same syntax for `refine` config: 08eb257a71248129f5f98eff202a68bb213e941f.

jameshadfield replied:
I'm biased but I really like the "option-first" syntax for refine (and other non-subsample rules) shown in 08eb257 etc

One minor note (not to be implemented here, just for reference), in avian-flu you can supply a scalar value directly if all wildcard combinations use the same value. E.g. the following are identical:
```yaml
  coalescent:
    '*/*/*': opt
  coalescent: opt
```

For subsample - reading the vaious YAMLs (and only _reading_ them, not actually testing them), I feel there's too much magic. I couldn't confidently tell you how each dataset will be filtered! I'm sure I could _learn_, but from a first impression it feels hard! To riff on various approaches, have you explored something like this:

```yaml
subsample:
  # Define which samples each dataset uses
  # (I couldn't work this out from just reading bff7e5 or e54f10, so contents may be wrong!)
  - dataset:
    '*/*/all-time':
      - all-time
    '*/*/(6y|3y)':
      - recent
      - background
  - samples:
    all-time:
      exclude: config/outliers_ppx.txt
      include:
        'a/*/*': config/include_a.txt
        'b/*/*': config/include_b.txt
      min_date: 1975-01-01
      max_sequences: 
        '*/*/*': 3000
        '*/F-antibody-escape/*': 2000
      ...
```

joverlee521 replied:
Yeah, I like the "option-first" approach in https://github.com/nextstrain/rsv/commit/bff7e53ba1691aff2a10d89681d9075326594dc2. I think it makes it easier to stitch together which options are used per build. However, I'm a little unsure about the concatenation of `query` and list options. This keeps the config concise, but makes it's more difficult to understand how the override of the default config works for these options.

At first glance, the glob syntax being a little daunting for users to grok and edit. They can read the `results/**_config.yaml` to see what the expanded config will be, but that does not help with editing or overriding the config...I wonder if we should _also_ support directly providing the subsample config, i.e.

```yaml
subsample:
  "a/genome/6y":
    samples:
      recent:
        ...
      background:
        ...
```
Then they can copy the expanded config into their workflow config and edit the values as needed.

victorlin replied:
Thanks both, let's stick with the option-first idea and make adjustments from there.

---

> [@jameshadfield] One minor note (not to be implemented here, just for reference), in avian-flu you can supply a scalar value directly if all wildcard combinations use the same value.

This works nicely for backwards compatibility. I'll look into adding it here.

---

> [@jameshadfield] For subsample… I feel there's too much magic.

If I understand correctly, the "magic" lies in the relationship between datasets and samples. As Jover said, `results/**_config.yaml` files should help, but yes it's not very obvious from the workflow config alone.

---

> [@jameshadfield] To riff on various approaches, have you explored something like this …

These are the differences I see in your example, let me know if I got anything wrong:

1. List instead of dict under `subsample`
2. A new declaration block
3. Updated config-defining syntax of samples > options > datasets

(1) is fine. Earlier I advocated for list with two reasons: (a) to have an intuitive merge order and (b) to avoid Snakemake's config merging issues. I've since realized that (a) is not much of an issue because dict keys are guaranteed to be ordered in Python. (b) is still an issue, but we're already using the separate key workaround elsewhere (`custom_subsample`, `additional_inputs`).

(2) seems reasonable. It's extra verbosity, but helps readability and could also be used for validation. We'd have to consider how it fits with the rest of the config. Something like this could work:

```yaml
subsample:
  samples:
    '*/*/(6y|3y)': [recent, background]
    # no need for all-time; config processing automatically creates a single sample
  options:
    query:
      '*/*/*': missing_data<1000
      …
```

(3) I'm not sure about this. The "samples > options > datasets" ordering feels backwards. Where would the global defaults go (`*/*/*`)?

---

> [@joverlee521] At first glance, the glob syntax being a little daunting for users to grok and edit...I wonder if we should also support directly providing the subsample config

Good idea. I suspect most users won't be producing very many datasets, and the direct subsample config will be easier to work with. I'll look into adding this support.

jameshadfield replied:
> let's stick with the option-first idea and make adjustments from there.

I find the subsampling section in bff7e53ba1691aff2a10d89681d9075326594dc2 much easier to read than e54f105fb9dcb8daa25443edec71751452708a3c. Although I further find the "samples > options > datasets" ordering more natural, as I tend to think of _samples_ as the most important concept.

> If I understand correctly, the "magic" lies in the relationship between datasets and samples. 

For me the magic is two-fold, although they're inter-related:
- not knowing which samples a dataset is using. In e54f105fb9dcb8daa25443edec71751452708a3c or bff7e53ba1691aff2a10d89681d9075326594dc2, do any datasets just use the default sample, or do they all use either `{all-time}` or `{recent, background}`? I _think_ bff7e53ba1691aff2a10d89681d9075326594dc2 is the former and e54f105fb9dcb8daa25443edec71751452708a3c is the latter. 
- What's the behaviour of the default sample? It seems to be both a sample on its own (for datasets which don't explicitly define sample names) and also merged into each and every explicitly defined sample?

(I know "default" is the word to use, but from the YAML structure it's more like the "implicit/unspecified" sample at the moment.)


> As Jover said, `results/**_config.yaml` files should help, but yes it's not very obvious from the workflow config alone.

I think we're never going to find a dead-simple config structure, but I think we can find one where you can figure it out by reading the YAML a couple of times, perhaps cross-referencing `results/**_config.yaml` to double-check. I think I understand the YAMLs in e54f105fb9dcb8daa25443edec71751452708a3c or bff7e53ba1691aff2a10d89681d9075326594dc2 but only after reading them many many times!

---

> (1) List instead of dict under subsample 

This is me being bad at YAML. 'dataset' and 'samples' were intended to be keys in the main 'subsample' dict, to better support merging. Merging is important in our configs more generally so we try to keep this possibility. That being said, the subsampling YAMLs are hard enough as they are, having easy-to-reason-with overlay YAMLs seems.... hard!

> (2) seems reasonable. It's extra verbosity, but helps readability and could also be used for validation. 

Making the association between datasets and samples clearer is a big one for me. If we go with the option-first idea something like your example in (2) is good, but... the "no need for all-time; config processing automatically creates a single sample" is more magic! We should find some way of explicitly saying that `'*/*/all-time'` uses the defaults. I came up with `samples['*/*/all-time'] = []`, but that's also a little magic, or `samples['*/*/all-time'] = ['DEFAULT']`, but that's odd because default doesn't appear in the defined samples for `*/*/(6y|3y)` even though it will be used.

> (3) ... Where would the global defaults go (`*/*/*`)?

I generally find the concept of defaults hard to reason with in all of these ideas. However I see that without it we have verbose YAMLs, even if we use anchors. (And of course you can write YAMLs which don't make any use of the defaults if you want to!)

I think clarity would be improved if we made it explicit. For bff7e53ba1691aff2a10d89681d9075326594dc2 this would look like:

```yaml
  options:
    include:
      DEFAULT:
        'a/*/*': config/include_a.txt
        'b/*/*': config/include_b.txt
```

For my "samples > options > datasets" configuration it would look like

```yaml
  samples:
    default: 
      include:
        'a/*/*': config/include_a.txt
        'b/*/*': config/include_b.txt
 ```

victorlin replied:
Agreed on options > datasets. However, I think treating samples as the most important concept makes defaults something that *needs* to be understood rather than implicit.

With "options > datasets > samples", configuration is applied in layers:

1. global defaults
2. dataset-level overrides
3. sample-level overrides

In this model, samples are an optional refinement. For some users, such as those using a single `augur filter` call, the default single-sample behavior is sufficient and they don't need to think about samples at all. When needed, datasets can then be split into multiple samples with more granular config values.

We can extend this idea to datasets as well. This works nicely with the avian-flu idea that options can take scalar values applied to all datasets.

Using a new example, let's look at the [current filter config for zika](https://github.com/nextstrain/zika/blob/7d90f92e3a86e330b2d591ca32bfb26d603be4f8/phylogenetic/defaults/config.yaml#L18-L25). With the single-dataset default behavior, swapping to `augur subsample` would be as simple as swapping the key name from `filter` to `subsample`.

```yaml
subsample:
  group_by:
    - country
    - year
    - month
  sequences_per_group: 40
  min_date: 2012
  min_length: 5385
```

When the need arises to start making multiple datasets (e.g. genome vs E gene), only dataset-specific values need to change. (unchanged values shown as comments)

```yaml
subsample:
#   group_by:
#     - country
#     - year
#     - month
#   sequences_per_group: 40
#   min_date: 2012
  min_length:
    'genome': 5385
    'E': 1400
```

When the need arises to start making multiple samples within a dataset (e.g. time resolutions), sample-level overrides can be introduced.

```yaml
subsample:
#   group_by:
#     - country
#     - year
#     - month
  max_sequences:
    '*/all-time': 1000
    '*/1y':
      early: 200
      recent: 800
  min_date:
    '*/all-time': 2012
    '*/1y':
      early: 2012
      recent: 1Y
  max_date:
    '*/1y':
      early: 1Y
#   min_length:
#     'genome/*': 5385
#     'E/*': 1400
```

In all cases, shared configuration remains unchanged unless explicitly overridden.

victorlin replied:
I've incorporated some of the ideas above into the latest changes.

> [@jameshadfield] Making the association between datasets and samples clearer is a big one for me.

Added in a28275d4080648ecfd599b0c2eee743453a4546d.

> [@jameshadfield] in avian-flu you can supply a scalar value directly if all wildcard combinations use the same value.

Done. Now the YAML diff in 852ebb705a1cfd76971900ba70b63e181ff5416d is easier to compare with the previous `filter` config.

> [@joverlee521] I'm a little unsure about the concatenation of `query` and list options.

Removed in 487e0f497199eab6abc4eb379f8617bce42bf77b.

> [@joverlee521] I wonder if we should also support directly providing the subsample config

Done using a separate key, mutually exclusive with `custom_subsample`:

https://github.com/nextstrain/rsv/blob/a28275d4080648ecfd599b0c2eee743453a4546d/README.md#L170-L183

[Preview of WA DOH config using `subsample_datasets`](https://github.com/nextstrain/rsv/blob/0d1ae93d0d246d64663a25263f04b8eed077e927/profiles/wadoh/configfile_wadoh.yaml#L5-L69)

joverlee521 replied:
Thanks for the latest iteration Victor! I'm pretty happy with where this has landed, but the goal post moves again...Now I'm questioning where would proximal subsampling parameters fit into the wildcard config? These are very sample focused so I don't see them as top level options next to `group_by`, `min_length`, etc. Would these nest under `samples`? 

```yaml
subsample:
  samples:
    '*/*/(6y|3y)': 
      - recent
      - background:
           focal_sample: recent
           k: 5
           max_distance: 5
```

victorlin replied:
Indeed, all the prototypes so far have not considered the more sophisticated config schema that is now supported by `augur subsample` with support for proximal samples and samples dependent on other samples.

`subsample_datasets` is an easy way out but we shouldn't rely on that. For the default config, the option-first structure clashes with the sample-first nature of the `augur subsample` config (now I see why James treats samples as the most important concept).

I don't think we should support both `subsample.<option>...` and `subsample.samples...<option>`. That would introduce yet another layer of merging to complicate things.

The option-first structure (option > datasets > samples) works best for filter-sample-only config and worst for config with proximal samples, while a sample-first structure (as James proposed) is the opposite. I wonder if reverting to a dataset-first structure would be a good middle ground.

jameshadfield replied:
(Apologies for the delayed reply - let's discuss this at a dev-chat meeting soon based on the gdoc you're creating! I'm glad the concept of tiered/hierarchical subsampling has entered this discussion, our config approach should support it.)

victorlin replied:
I've dropped the extended glob syntax from this PR. The latest iteration of that can be found in https://github.com/nextstrain/measles/pull/131.

This PR now configures `augur subsample` the same way as other repos. That's 24 per-build configs, so I've added a script to generate the entire config file.

victorlin committed:
Move constants up in file

Preparing to use build_dir in config.smk, and I figured it'd be good to
move both auspice_dir with it.

victorlin committed:
Use augur subsample

The previous subsampling implementation was fixed to a two-sample
recent+background split with some hardcoded parameters. Replacing it
with augur subsample allows for more flexible configuration.

To keep the workflow config schema concise, we generate each augur
subsample config dynamically using a patterns defined in the config
helper functions in config.smk.

This is a breaking change and the old configuration will no longer work.

victorlin committed:
Add separate config section for filter_for_pre_subsample_alignment

Similar to "Add separate frequencies config" (0b221851), this rule
shouldn't rely on config from another rule.

victorlin committed:
Define valid dataset parts

To be used for validating augur subsample config.

victorlin committed:
🚧 option-first glob syntax

victorlin committed:
🚧 glob syntax for refine

victorlin committed:
Use augur subsample

The previous subsampling implementation was fixed to a two-sample
recent+background split with some hardcoded parameters. Replacing it
with augur subsample allows for more flexible configuration.

In Snakemake, implementation is mostly copied from pathogen repos that
have switched over to augur subsample. One notable difference is that
the combine_samples rule must stay to handle output from the
enrich_antibody_escape rule.

In the config YAML, the subsampling configuration is much more verbose
as a byproduct of increased flexibility. It was generated using a
script, which I'll add in another commit since it makes additional
changes.

victorlin committed:
Generate default config YAML

This makes it easier to make changes to the subsample config.

One downside is that the generated file is less readable with a strict
YAML style and no comments. Comments have been moved to the script, but
ideally they'd live in a schema which is used to generate user-facing
docs.

victorlin committed:
Add separate config section for filter_for_pre_subsample_alignment

Similar to "Add separate frequencies config" (0b221851), the
filter_for_pre_subsample_alignment rule shouldn't rely on config from
another rule.

joverlee521 commented on scripts/generate_default_config.py:
Small request to update the [config section in the README](https://github.com/nextstrain/rsv/blob/master/README.md#configuration) to include instructions to edit/run this script to update the config.

joverlee521 commented on .gitattributes:
```diff
@@ -1,3 +1,8 @@
 # Always use LF line endings even on Windows.
 * text=auto eol=lf
+
+# This is a large generated file that, while text, it is not useful to
+# routinely show the diff of.  A diff can be forced as needed, e.g. with `git
+# diff --text`.
+/config/configfile.yaml -diff
```
Hmm, I actually think it'd be helpful to have the diff to confirm that script changes are generating the expected configfile.yaml. Other than this change, I'd think future edits should be pretty small diffs?

joverlee521 commented on config/configfile.yaml:
```diff
@@ -107,6 +107,25 @@ nextclade_attributes:
     reference_name: "hRSV/B/Australia/VIC-RCH056/2019"
     accession: "EPI_ISL_1653999"
 
+filter_for_pre_subsample_alignment:
```
_non-blocking_

I haven't kept up with development in RSV, so I wasn't sure what this was used for. Looking at the DAG, this rule/config is only used for the F-antibody-escape build. Should the param to something more specific (e.g. `filter_for_f_antibody_escape`) to prevent confusion?

joverlee521 commented on CHANGELOG.md:
```diff
@@ -4,4 +4,4 @@ We use this CHANGELOG to document breaking changes, new features, bug fixes, and
 ## 2026
 
-* TBD: Phylogenetic workflow configuration for initial quality filtering has moved to its own section, `filter_for_pre_subsample_alignment`.
+* TBD: The `filter` section in phylogenetic workflow configuration has been replaced by `subsample`/`custom_subsample` for subsampling, and `filter_for_pre_subsample_alignment` for initial quality filtering. **This is a breaking change**.
```
This is another instance where proximity subsampling is not supported so we should note it in the config/changelog (similar to [rubella](https://github.com/nextstrain/rubella/pull/62#discussion_r3716753387)). 

We're going to continuously run into this issue if we don't reorder workflows to align before subsample...I wonder if we should build a way to flag this in the subsample schema validation?

joverlee521 reviewed (approved):
Thank you for continuing to push through this! I've only added small comments for docs/changelog edits.

victorlin committed:
Rename rule to filter_for_f_antibody_escape

The new name makes it more obvious that this rule is only used when
build_name=F-antibody-escape.

rneher commented:
Thanks for pushing this, Victor. I am a bit on the fence here. What bugs me most about this is that if one now wants to change something in the workflow, one has to trace the config that ends up as input in the rules through even more layers of input functions and config generators. I am wondering whether part of this problem could be avoided if instead of using `write_subsample_config` to pull out a specific section of the main file, we just use the `generate_default_config` to produce that section and place it as `subsample_config` into the corresponding results directory? This would also avoid having to commit the generated config and the resulting duplication parameters. The config could then specify custom subsampling configs if desired and fall back on the autogenerated ones.

victorlin commented:
Users currently rely on values in the default config file for inheritance and documentation. There are better ways to provide documentation, but values in the default config file is necessary for the inheritance aspect. So I don't think we can save the generated default subsample values directly to results without breaking config inheritance.

That said, I take your point on increased complexity in how config is passed around, though for this PR I think that comes more from the switch to `augur subsample` – the pattern of providing config via an exported YAML file – than from the switch to generating default config via script.

I can drop c33cf9220b5a80a7546f5e09f1a798b176852ab4 if you/others are ok with hand-editing the default config file, similar to how things are done with ncov's [builds.yaml](https://github.com/nextstrain/ncov/blob/b4ee8292d1a8c73031d3057abf80d186faf10a47/nextstrain_profiles/nextstrain-open/builds.yaml).
